src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+                case ConversionKind.Identity:           // a value of a given type matches a pattern of that type
+                case ConversionKind.ImplicitReference:  // a value of type string matches a pattern of type object
+
+                case ConversionKind.ImplicitNullable:   // a value of type int matches a pattern of type int?
+                case ConversionKind.ExplicitNullable:   // a non-null value of type "int?" matches a pattern of type int
+                    // but if the types differ (e.g. one of them is type byte and the other is type int?).. no match
+                    return expressionType.StrippedType().TupleUnderlyingType == patternType.StrippedType().TupleUnderlyingType;
+
+                case ConversionKind.ExplicitEnumeration:// a value of enum type does not match a pattern of integral type
+                case ConversionKind.ExplicitNumeric:    // a value of type long does not match a pattern of type int
+                case ConversionKind.ImplicitNumeric:    // a value of type short does not match a pattern of type int
+                case ConversionKind.ImplicitTuple:      // distinct tuple types don't match
+                case ConversionKind.NoConversion:
+                    return false;
+
+                case ConversionKind.ExplicitDynamic:    // a value of type dynamic might not match a pattern of type other than object
+                case ConversionKind.ExplicitReference:  // a narrowing reference conversion might or might not succeed
+                case ConversionKind.Unboxing:           // a value of type object might match a pattern of type int
+                    return null;
+
+                default:
+                    // other conversions don't apply (e.g. conversions from expression, user-defined) and should not arise
+                    throw ExceptionUtilities.UnexpectedValue(conversion.Kind);
+            }
+        }
+    }
+
+    internal abstract class DecisionTree
+    {
+        public DecisionTree Parent;
+        public readonly BoundExpression Expression;
+        public readonly TypeSymbol Type;
+        public bool MatchIsComplete;
+
+        public enum DecisionKind { ByType, ByValue, Guarded }
+        public abstract DecisionKind Kind { get; }
+
+        public DecisionTree(BoundExpression expression, TypeSymbol type)
+        {
+            this.Expression = expression;
+            this.Type = type;
+            Debug.Assert(this.Expression != null);
+            Debug.Assert(this.Type != null);
+        }
+
+        public static DecisionTree Create(BoundExpression expression, TypeSymbol type)
+        {
+            if (expression.Type.CanBeAssignedNull())
+            {
+                // we need the ByType decision tree to separate null from non-null values
+                return new ByType(expression, type);
+            }
+            else
+            {
+                // If it is a builtin value type, we can switch on its (constant) values
```

Comment:
We don't need to. I'll add a clarifying comment.

---
In reply to: [64983982](https://github.com/dotnet/roslyn/pull/11485#discussion_r64983982) [](ancestors = 64983982)